### DISCUSSION TITLE
Using attribute names in UPDATE.

### DISCUSSION THREAD

**Discussion by Dev03Ibraflex** [psycopg/psycopg#880] [Q&A] [answered]
Hello! I'm trying to use two tuples to perform an UPDATE in the following format:

```
        query = pgsql.SQL('UPDATE {table} SET {columns} = {values} WHERE name = {name} RETURNING name').format(
            table = pgsql.Identifier(ProfileModel.__tablename__),
            columns = tuple(f for f in new_profile.__dict__.keys()), 
            values = tuple(f for f in new_profile.__dict__.values()), 
            name = name_profile)

        print(query.as_string())
```

However, this is generating invalid SQL for me:

`UPDATE "profiles" SET '(name,description,feats)' = '(HomemMorcego,"Faz tudo agora","{t,t,t}")' WHERE name = 'Batman' RETURNING name`

```
syntax error at or near "'(name,description,feats)'"
LINE 1: UPDATE "profiles" SET '(name,description,feats)' = '(HomemMo...
```

How could I generate an SQL in the format:

```
-- Only example
UPDATE anytable SET (col1, col2, col3) = (1, 'anyval', True)
  WHERE col0 = 'key';
```

Without having to mention the properties of the `ProfileModel` object one by one?

**Comment by dvarrazzo** [comment]
What is a syntax that works in psql?

**Reply by Dev03Ibraflex** [reply]
`UPDATE "profiles" SET (name, description, feats) = ('HomemMorcego','Faz tudo agora','{t,t,t}') WHERE name = 'Batman' RETURNING name;`

Values ​​must have single quotes. And parentheses cannot be between quotation marks. Eg.

`UPDATE "profiles" SET '(name,description,feats)' = '(HomemMorcego,"Faz tudo agora","{t,t,t}")' WHERE name = 'Batman' RETURNING name`.

              This is a error ^

**Reply by dvarrazzo** [reply, marked as answer]
As far as I can see, you should wrap every single column name in a `sql.Identifier` and every single value in a `sql.Literal`. You should join these sequences using `sql.SQL(",")`.

**Reply by Dev03Ibraflex** [reply]
Yes, here is a example:

```python
query = pgsql.SQL(
    "UPDATE {table} SET ({columns}) = ({values}) WHERE id = {id}").format(
        table=pgsql.Identifier(BoxModel.__tablename__),
        columns=pgsql.SQL(", ").join([pgsql.Identifier(k) for k in box_model.__dict__.keys()]),
        values=pgsql.SQL(", ").join([v for v in box_model.__dict__.values()]),
        id=box_model.id
    )
```
..

**Reply by dvarrazzo** [reply]
Looks good. You might even create a query with [placeholders](https://www.psycopg.org/psycopg3/docs/api/sql.html#psycopg.sql.Placeholder) and then pass the values at `execute()` time. This would have the advantage that you would need to compose your query only once, and, because the query string doesn't change, psycopg would make better use of prepared statements. Untested...

```python
query = pgsql.SQL(
    "UPDATE {table} SET ({columns}) = ({placeholders}) WHERE id = %(id)s").format(
        table=pgsql.Identifier(BoxModel.__tablename__),
        columns=pgsql.SQL(", ").join([pgsql.Identifier(k) for k in box_model.__dict__.keys()]),
        placeholders=pgsql.SQL(", ").join([pgsql.Placeholder(k) for k in box_model.__dict__.keys()]),
    )

# ...
cursor.execute(query, box_mode.__dict__)
```
